File: leetcode/python/hard/4_median_of_two_sorted_arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        m, n = len(nums1), len(nums2)
        if (m+n) & 1:
            return self.findKth(nums1, nums2, (m+n)//2)
        else:
            return (self.findKth(nums1, nums2, (m+n)//2) + self.findKth(nums1, nums2, (m+n)//2-1)) / 2
    
    def findKth(self, nums1: list[int], nums2: list[int], k: int) -> int:
        if not nums1:
            return nums2[k]
        if not nums2:
            return nums1[k]
        
        m1, m2 = len(nums1)//2, len(nums2)//2
        if m1 + m2 < k: # need more elements
            if nums1[m1] > nums2[m2]:
                return self.findKth(nums1, nums2[m2+1:], k-m2-1)
            else:
                return self.findKth(nums1[m1+1:], nums2, k-m1-1)
        else:
            if nums1[m1] > nums2[m2]:
                return self.findKth(nums1[:m1], nums2, k)
            else:
                return self.findKth(nums1, nums2[:m2], k)

s = Solution()
print(s.findMedianSortedArrays(nums1 = [1,3], nums2 = [2]))
print(s.findMedianSortedArrays(nums1 = [1,2], nums2 = [3,4]))

File: leetcode/rust/src/hard/4_median_of_two_sorted_arrays.rs
impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let m = nums1.len();
        let n = nums2.len();

        if (m + n) % 2 == 0 {
            return (Self::find_kth(&nums1, &nums2, (m + n) / 2) + Self::find_kth(&nums1, &nums2, (m + n) / 2 - 1)) as f64 / 2.0;
        } else {
            return Self::find_kth(nums1.as_slice(), nums2.as_slice(), (m + n) / 2) as f64;
        }
    }

    fn find_kth(nums1: &[i32], nums2:  &[i32], k: usize) -> i32 {
        if nums1.is_empty() {
            return nums2[k];
        }
        if nums2.is_empty() {
            return nums1[k];
        }

        let m = nums1.len() / 2;
        let n = nums2.len() / 2;
        if m + n < k {
            if nums1[m] > nums2[n] {
                return Self::find_kth(nums1, &nums2[n + 1..], k - n - 1);
            }
            else {
                return Self::find_kth(&nums1[m + 1..], nums2, k - m - 1);
            }
        }
        else {
            if nums1[m] > nums2[n] {
                return Self::find_kth(&nums1[..m], nums2, k);
            }
            else {
                return Self::find_kth(nums1, &nums2[..n], k);
            }
        }

    }
}